simba/costs.py

```diff
@@ -1,260 +1,578 @@
+            "c_vehicles_annual", "c_gcs_annual", "c_cs_annual", "c_garage_annual",
+            "c_stat_storage_annual", "c_feed_in_annual", "c_invest_annual",
+            # annual maintenance costs
+            "c_maint_vehicles_annual", "c_maint_gc_annual", "c_maint_cs_annual",
+            "c_maint_stat_storage_annual", "c_maint_feed_in_annual",
+            "c_maint_infrastructure_annual",
+            "c_maint_annual",
+            # annual electricity costs
+            "c_el_procurement_annual", "c_el_power_price_annual",
+            "c_el_energy_price_annual",
+            "c_el_taxes_annual", "c_el_feed_in_remuneration_annual", "c_el_annual"]
+
+    def get_columns(self):
+        """
+        Get the sorted columns for cost calculations.
+
+        :return: The list of columns.
+        :rtype: list
+        """
+        return list(dict(sorted(self.costs_per_gc.items(), key=lambda x: x[1][self.SORT_COLUMN],
+                                reverse=True)).keys())
+
+    def get_gcs(self):
+        """
+        Get the grid connectors used in the scenario.
+
+        :return: The grid connectors.
+        :rtype: dict
+        """
+        return self.scenario.components.grid_connectors
+
+    def to_csv_lists(self):
+        """
+        Convert costs to a list of lists easily converteable to a csv.
+
+        :return: The list of lists of parameters, units and costs per gc.
+        :rtype: list
+        """
+        output = [["parameter", "unit"] + self.get_columns()]
+        for key in self.get_vehicle_types():
+            row = [key, self.get_unit(key)]
+            for col in self.get_columns():
+                row.append(self.vehicles_per_gc[col][key])
+            output.append(row)
+
+        for key in next(iter(self.costs_per_gc.values())):
+            row = [key, self.get_unit(key)]
+            for col in self.get_columns():
+                row.append(round(self.costs_per_gc[col][key], self.rounding_precision))
+            output.append(row)
+        return output
+
+    def get_vehicle_types(self):
+        """
+         Get the types of vehicles used.
+
+         :return: The vehicle types.
+         :rtype: list
+         """
+        return next(iter(self.vehicles_per_gc.values())).keys()
+
+    def cumulate(self):
+        """
+        Cumulate the costs of vehicles and infrastructure.
+
+        :return: The updated instance.
+        :rtype: CostsObject
+        """
+        v_types = self.get_vehicle_types()
+        self.vehicles_per_gc[self.CUMULATED] = dict()
+        for v_type in v_types:
+            self.vehicles_per_gc[self.CUMULATED][v_type] = sum(
+                [self.vehicles_per_gc[gc][v_type] for gc in self.get_gcs()])
+
+        # Cumulate gcs variables
+        self.costs_per_gc[self.CUMULATED] = dict()
+        for key in next(iter(self.costs_per_gc.values())):
+            self.costs_per_gc[self.CUMULATED][key] = 0
+            for gc in self.costs_per_gc.keys()-[self.CUMULATED]:
```

Comment:
I replaced gcs() with an attribute of the same name and included a variable gcs_and_garage to use for iteration over both. Vehicle counting should not use get(... , 0) imo, since the function call requires precalculated vehicle counts of each grid connector. A missing value is an indication of wrong usage or faulty code. Both cases should not be handled by giving default values